### PR TITLE
Feature SkyTruth Alerts map as a layer in Leaflet profile map #2427

### LINKED ISSUE #2427: Feature SkyTruth Alerts map as a layer in Leaflet profile map
We have a small map that appears on some profile pages like: http://publiclab.org/profile/nshapiro

We should create a Leaflet map layer that can be disabled, showing nearby SkyTruth Alerts; see:

* [x] create a new layer that can be enabled/disabled with http://leafletjs.com/examples/layers-control/
* [x] add a custom icon with http://leafletjs.com/examples/custom-icons/
* [x] make new markers with `L.marker([51.5, -0.09], {icon: greenIcon}).addTo(map).bindPopup("I am a green leaf.");`

Example JSON file for a longitude/latitude box: http://alerts.skytruth.org/json?&d=30&channel=local&l=41.545%2C-71.8331%2C41.8782%2C-71.3524

Here's a broader area: http://alerts.skytruth.org/json?&d=30&channel=local&l=41.042%2C-72.554%2C42.3747%2C-70.6314

Each alert has a good description that could be shown in a popup, with a link to more information:

```json
{ "feed":[
{
    "id": "4cb6a3cf-7db5-314a-9340-ebe44628fdc1",
    "source": "1",
    "title": "NRC Report: Caustic Gas near Cranston, RI",
    "link": "http://alerts.skytruth.org/report/4cb6a3cf-7db5-314a-9340-ebe44628fdc1",
    "description": "<b>Report Details<\/b><br\/>NRC Report ID: <a href=\"http:\/\/nrc.uscg.mil\/\">1203212<\/a><br\/>Incident Time: 2018-01-29 04:30:00<br\/>Nearest City: Cranston, RI<br\/>Location: 1420 ELMWOOD AVE<br\/>Incident Type: FIXED<br\/>Material: CAUSTIC GAS<br\/>Medium Affected: LAND<br\/>Suspected Responsible Party: <br\/><b>SkyTruth Analysis<\/b><br\/>Lat\/Long: 41.770898, -71.422979 (Approximated from premise)<br\/><b>Report Description<\/b><br\/>CALLER STATED THERE WAS A RELEASE OF CAUSTIC GAS FROM A BUILDING DUE TO A SMALL FIRE. CALLER ALSO STATED THERE WERE ADDITIONAL UNKNOWN CHEMICALS THAT WERE RELEASED. CALLER STATED THE REASON FOR THE RELEASE IS UNKNOWN AT THIS TIME. CALLER STATED THIS IS A PLATING COMPANY.<p><a href=\"http:\/\/alerts.skytruth.org\/report\/4cb6a3cf-7db5-314a-9340-ebe44628fdc1\" target=\"_top\">View Full Report<\/a><\/p>",
    "lat": "41.7708981",
    "lng": "-71.4229789",
    "published": "2018-02-14 05:50:22.009",
    "updated": "2018-02-14 05:50:22.009",
    "incident_datetime": "2018-01-29 04:30:00"    
}

]
}
```

The code for the user profile map is here: 

https://github.com/publiclab/plots2/blob/2d43c7d908dd8b65d3ca7681cc35a107770e0f20/app/views/map/_userLeaflet.html.erb#L8-L12

Let's be sure to highlight SkyTruth Alerts in the map key! http://alerts.skytruth.org

### PR DESCRIPTION
**Pull request by sagarpreet-chadha**
Fixes #2427 .
To-Do : 
* [x] Add different markers . 
* [x] Way to expand the map .  

1.) Currently fetching only 10 records .
2.) Fetching only those Records near my location . 

Thanks!

### CODE REVIEW

**Comment by sagarpreet-chadha** [comment]
@jywarren ...can you provide me the link of the leaflet-marker image and its shadow from [here](http://leafletjs.com/examples/custom-icons/) . Thanks !

**Comment by sagarpreet-chadha** [comment]
gif

**Comment by sagarpreet-chadha** [comment]
Re-Triggering

**Comment by jywarren** [comment]
Oh wow!!!!!! This is amazing!

Shall we look for a nice small icon from fontawesome that we can use for
the marker?

Fantastic work!

**Comment by sagarpreet-chadha** [comment]
Thanks so much ! 
FontAwesome requires us to give a license link if we use their icons . 

Further we can add more **different color** markers for more layers (like **Near_by Projects** , etc.) . What do you think about this ? 

screen shot 2018-03-11 at 5 06 06 pm

**Comment by jywarren** [comment]
Sure - but don't we have a fontawesome credit on the footer of the site? We
can add one there if we like. Another color also sounds great.

**Comment by sagarpreet-chadha** [comment]
Some options : 

screen shot 2018-03-11 at 11 42 07 pm

screen shot 2018-03-11 at 11 41 29 pm

screen shot 2018-03-11 at 11 40 22 pm

screen shot 2018-03-11 at 11 17 30 pm

screen shot 2018-03-11 at 11 11 23 pm

screen shot 2018-03-11 at 5 06 06 pm

**Comment by jywarren** [comment]
I guess I'm thinking something quite small. See how on http://alerts.skytruth.org, they appear as small red dots? Could we do something like that?

Thanks @sagarpreet-chadha for taking time to get this just right! It's very exciting!

**Comment by sagarpreet-chadha** [comment]
Thanks @jywarren  ! How about this one ?
screen shot 2018-03-13 at 1 23 42 pm

The link of Dot marker is   [this](http://www.clker.com/cliparts/T/G/b/7/r/A/red-dot.svg) .

One thing to note here is that when two dots are very close to each other , we need to zoom a little so that their corresponding pop-ups appear .  This happens in [SkyTruth Alert Maps](http://alerts.skytruth.org/) too .

**Comment by sagarpreet-chadha** [comment]
For Full - Screen , i found this plugin which opens the map on **fullscreen**  via a button : 

fullscreen_gif

But the full-screen kind of makes the map confusing because we see repeated continents . 
I think this #2248 issue may solve this on Full-Screen .  What do you think @jywarren , @ebarry  ?

**Comment by sagarpreet-chadha** [comment]
Some Ideas : 

1.) If we choose to use Full-Screen map , we can show **all** the markers and not only which are near the user . 

2.) Also on SkyTruth Alerts map [here](http://alerts.skytruth.org/) ... i noticed that when we **zoom - in** , the number of markers on map **increases**  and vice-versa . We can also implement this for all leaflet maps that we use . I think this is related to #1934 . 

Thanks !

**Comment by jywarren** [comment]
Let's do this in 2 steps-- i LOVE the dots on your last mockup. For full-screen, I think we can just add that button now, and try to optimize it more later. 

One follow-up could be to zoom in a bit when you go full screen. 

This is awesome!!!!!! 🎉

**Comment by ebarry** [comment]
Wow this is cool! Quick question because the gif shows the map in a profile page, but in my opinion at this time, these layers are not appropriate for maps on profile pages. Where would a map of environmental layers be on publiclab.org? thanks for helping me get my head around this!

**Comment by sagarpreet-chadha** [comment]
Hi @ebarry ! 
I think as discussed here :  [proposal on PublicLab.org](https://publiclab.org/notes/sagarpreet/02-16-2018/social-media-integration-and-part-of-interactive-project-maps#c19032) , we can make a separate JS library for such maps and then it would be easy to show a Map with **required features at appropriate places** on our website . This JS library will be used for all maps like **/people** , **/profile** ,etc. maps . 
What do you think ? Thanks !

**Comment by jywarren** [comment]
Yes, i think this sounds good -- to develop in a minimal way on the profile page now, with a plan to roll this out on other maps across the site. @ebarry if you have a suggested location to display these layers, please chime in! 

For now, the profile page is the only place you see "where you are" so it seems appropriate to begin with that, and it's pretty unobtrusive at this point. 

Later if you want to propose another page that shows "where you are" we can add a link from the profile map to that page. But consider:

1. what the URL would be
2. what the page title would be
3. what else that page might say

Then we'll have a way forward on that too. Thanks, Liz!

**Comment by jywarren** [comment]
Is this otherwise ready for merging, though? I'm restarting the failed test run, i think it was an unrelated issue since i don't see the other one failed too.

**Comment by sagarpreet-chadha** [comment]
Yes @jywarren , it is ready for merging .
@publiclab/reviewers team...kindly review this :)

**Comment by jywarren** [comment]
This is so odd. What do you think the issue is here, @publiclab/reviewers ??

```
W: Failed to fetch ftp://ftp.us.debian.org/debian/dists/jessie/Release.gpg  Unable to fetch file, server said 'Security: Bad IP connecting.  ' [IP: 128.61.240.89 21]
W: Failed to fetch ftp://ftp.us.debian.org/debian/dists/jessie-updates/Release.gpg  Unable to fetch file, server said 'Security: Bad IP connecting.  ' [IP: 128.61.240.89 21]
W: Failed to fetch ftp://ftp.us.debian.org/debian/dists/jessie/main/binary-amd64/Packages  Unable to fetch file, server said 'Failed to open file.  ' [IP: 128.61.240.89 21]
W: Failed to fetch ftp://ftp.us.debian.org/debian/dists/jessie-updates/main/binary-amd64/Packages  Unable to fetch file, server said 'Failed to open file.  ' [IP: 128.61.240.89 21]
E: Some index files failed to download. They have been ignored, or old ones used instead.
Service 'web' failed to build: The command '/bin/sh -c apt-get update -qq && apt-get install -y bundler libmysqlclient-dev ruby-rmagick libfreeimage3 nodejs-legacy npm wget openjdk-7-jre openjdk-7-jdk procps' returned a non-zero code: 100
```

**Comment by Gauravano** [comment]
I guess above error relates to #2471 , what do you guys think?

**Comment by jywarren** [comment]
Hm, do you think? Has it been rebased over the fix I submitted?

**Comment by Gauravano** [comment]
I don't think it has rebased as in that case other builds may have  failed too

File: app/views/map/_userLeaflet.html.erb [hunk 1]
```diff
@@ -4,12 +4,49 @@
...
+    }
+    var baseMaps = {
+      "Default": normal_layer,
+    };
+    var overlayMaps = {
+      "Skynet": layerGroup    // we can add more layers here !
+    }; 
+    L.control.layers(baseMaps , overlayMaps).addTo(map<%= unique_id %>);
+    <% if !(@map_blurred == true) %>
+      L.marker([<%= @map_lat %>, <%= @map_lon %>]).addTo(map<%= unique_id %>).bindPopup("<a href='https://publiclab.org/profile/<%= @profile_user.username  %>'><%= @profile_user.username  %> </a>");
```
Comment (Gauravano) [inline app/views/map/_userLeaflet.html.erb]: Script is really nice !!!

**Comment by jywarren** [comment]
We actually have to get this to pass before merging. Maybe adding a extra commit would clear this out somehow?

**Comment by Gauravano** [comment]
@jywarren error is generated by travis-ci/push . Is it possible if @sagarpreet-chadha has opened this PR when we were facing https://github.com/publiclab/plots2/issues/2471 issue and as travis-ci/push doesn't re-builds on closing and opening the PR

**Comment by Gauravano** [comment]
In your PR https://github.com/publiclab/plots2/pull/2459, same error is there on travis-ci/push

**Comment by sagarpreet-chadha** [comment]
Okay , trying to figure out the problem here 👍 .

**Comment by jywarren** [comment]
I think perhaps it's worth trying to open a new PR -- that worked for another similar issue... Just be sure to link from here!

**Comment by sagarpreet-chadha** [comment]
Shifted to #2536 .

**Comment by sagarpreet-chadha** [comment]
screen shot 2018-03-24 at 4 09 36 pm

**Pull request closed without merging**